### ISSUE TITLE
string.replace() should allow taking a function as second argument?

### ISSUE THREAD

**Issue by user202729** [typst/typst#393]
As in the title. (similar to how Python allow a lambda as the replacement in `re.sub`, which will be applied on match objects)

**Comment by dvdvgt** [comment]
I would like to work on this. Could you perhaps provide a short example on how you would expect this to function? More specifically, whether this should be a breaking change where the [`replacement` parameter](https://typst.app/docs/reference/types/string/#replace--parameters--replacement) should now be of type `function` instead of `string`.

I think changing the type is sensible because you can always just pass a function that ignores its argument if you just want a simple replacement.

## Example

```
"123".replace(regex("\d+"), "abc")
```

becomes

```
"123".replace(regex("\d+"), _ => "abc")
```

**Comment by laurmaedje** [comment]
both should be supported, string or function

**Comment by laurmaedje** [comment]
Fixed via #944.
